python/cudf/cudf/core/dataframe.py

```diff
@@ -5040,82 +5041,83 @@ def isin(self, values):
+
+                    if self_is_cat != other_is_cat:
+                        # It is valid to compare the levels of a categorical
+                        # column to a non-categorical column.
+                        if self_is_cat:
+                            self_col = self_col._get_decategorized_column()
+                        else:
+                            other_col = other_col._get_decategorized_column()
 
-            result = DataFrame()
-            # TODO: propagate nulls through isin
-            # https://github.com/rapidsai/cudf/issues/7556
-            for col in self._data.names:
-                if isinstance(
-                    self[col]._column, cudf.core.column.CategoricalColumn
-                ) and isinstance(
-                    values._column, cudf.core.column.CategoricalColumn
-                ):
-                    res = (self._data[col] == values._column).fillna(False)
-                    result[col] = res
-                elif (
-                    isinstance(
-                        self[col]._column, cudf.core.column.CategoricalColumn
-                    )
-                    or np.issubdtype(self[col].dtype, cudf.dtype("object"))
-                ) or (
-                    isinstance(
-                        values._column, cudf.core.column.CategoricalColumn
-                    )
-                    or np.issubdtype(values.dtype, cudf.dtype("object"))
-                ):
-                    result[col] = utils.scalar_broadcast_to(False, len(self))
-                else:
-                    result[col] = (self._data[col] == values._column).fillna(
-                        False
-                    )
-
-            result.index = self.index
-            return result
-        elif isinstance(values, DataFrame):
-            values = values.reindex(self.index)
+                    # We use the type checks from _before_ the conversion
+                    # because if only one was categorical then it's already
+                    # been converted and we have to check if they're strings.
+                    if self_is_cat and other_is_cat:
+                        self_is_obj = other_is_obj = False
+                    else:
+                        # These checks must happen after the conversions above
+                        # since numpy can't handle categorical dtypes.
+                        self_is_obj = np.issubdtype(self_col.dtype, obj_dtype)
+                        other_is_obj = np.issubdtype(
+                            other_col.dtype, obj_dtype
+                        )
 
-            result = DataFrame()
-            for col in self._data.names:
-                if col in values.columns:
-                    result[col] = (
-                        self._data[col] == values[col]._column
-                    ).fillna(False)
+                    if self_is_obj != other_is_obj:
+                        # Strings can't compare to anything else.
```

Comment:
This is one of the edge cases I think. We do not have to address it here, as it was in the original implementation as well I believe. But technically string can compare as equal to 'string category':

```python
> pd.Series(['a','b','c'], dtype='category').isin(['a'])
0     True
1    False
2    False
dtype: bool
```

perhaps just an update to this comment or a todo would suffice.

Reply:
Actually I'm pretty sure this is precisely one of the cases that this PR fixes. If you look at the logic above, categorical columns are decategorized before we get here. The purpose of that was to make it so that by the time we define `self_is_str` and `other_is_str` we're looking at the underlying types of categories. I'm waiting for this branch to recompile for me and then I'll verify, but I think this case should work correctly after this PR.

Reply:
Yep this is fixed now.

```
>>> s = pd.Series(['a','b','c'], dtype='category')
>>> s.isin(['a'])
0     True
1    False
2    False
dtype: bool
>>> cudf.from_pandas(s).isin(['a'])
0     True
1    False
2    False
dtype: bool
```

The comment is referring to cases where you try to do 
```
>>> s = pd.Series(['a','b','c'])
>>> s.isin([1])
0    False
1    False
2    False
dtype: bool
>>> cudf.from_pandas(s).isin([1])
0    False
1    False
2    False
dtype: bool
```